### PR TITLE
ETH reimbursement pool for RFC-6

### PR DESCRIPTION
**Pull request by dimpar**
This PR implements [the ETH pool described in RFC-6](https://github.com/keep-network/tbtc-v2/blob/7b3cb19f6860cfc32677b15d5de643a9ee0f0f51/docs/rfc/rfc-6.adoc#32-eth-pool) and [this issue](https://github.com/keep-network/keep-core/issues/2824). ETH pool is used for transaction reimbursements executed on authorized contracts by the DAO.

### CODE REVIEW

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:110]: Should we limit this to DAO only?

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:30]: Could we initialize the values? Either assign them here or in a constructor.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:56]: Would this work?

```suggestion
        // slither-disable-next-line low-level-calls
        (bool sent, ) = receiver.call{value: refundAmount}("");
```

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:55]: Would it be worth emitting an event that a refund was made?

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:61]: Could we replace `governance` with `owner` across this contract?

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:57]: We need to remember to catch this error when we use this function, so we don't end up with a failing transaction. Maybe it would be worth adding a note about it in `@dev` docs?

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:107]: Can the disable be simplified?
```suggestion
        // slither-disable-next-line low-level-calls
        (bool sent, ) = receiver.call{value: amount}("");
```

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:123]: Do we really need this function? If the answer is yes, maybe we could simply call the `withdraw`?

```suggestion
    function withdrawAll(address receiver) external onlyOwner {
        withdraw(address(this).balance, receiver);
    }
```

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:126]: IMHO it's good this way.

**Review by nkuba — CHANGES_REQUESTED** [review]

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:55]: Not sure if it's worth it given we need to keep in mind the extra gas cost of the wrapping code. If we ever need any stats, we can simply find all out-transfers from this contract.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:25]: Not only the TX, also all the wrapping code that must be executed for the refund.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:123]: I think we need it so that we do not have to race to call `withdraw(uint256 amount, address receiver)` with transactions using the pool. See that it fails if the amount does not match.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:34]: Small note, we could name them `newStaticGas` and `newMaxGasPrice` to clearly state these are the new values.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:57]: I would ignore the result of `.call` in this function. For externally owned accounts `call` should always work. If the caller is a smart contract, it's that smart contract's responsibility to make sure Ether can be received. If there is some problem with receiving Ether, we do not need to revert, especially since we would revert at the very end, when the most complex and expensive logic has been already executed and paid for before `refund` is called. If we go this way, let's make sure to explain it in `@dev` section of docs.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:126]: If someone wants to donate, there is no reason for stopping them.

**Review by pdyraga — COMMENTED** [review]
Was looking only at `*.sol`.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:25]: Actually, thinking a bit more about it, do we need to make static gas var updateable? The cost of the transaction is constant which is 21000 & the cost of the entire refund function is around 16500. Total 37500. Gas amount will stay the same unless we change the refund function in a new contract, where we'll need to adjust the gas costs accordingly. So I'm thinking of making only the `maxGasPrice` updatable.

Comment (pdyraga) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:25]: Opcode cost can change, an example is https://eips.ethereum.org/EIPS/eip-2929. If we want to be able to adjust without redeploying `ReimbursmentPool`, we need to have this value updatable.

Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:25]: Ah, in that case, it makes sense to make a static gas cost updatable. 👍

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:32]: Yup, good idea [3138b17](https://github.com/keep-network/keep-core/pull/2826/commits/3138b1714cfab6d81ce0403c4b1b50d987fe73c3)

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: I tried, but it didn't

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: This is a refund function that will be usually called in already heavy gas-wise functions, so adding extra should be probably avoided.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: [11dc5a9](https://github.com/keep-network/keep-core/pull/2826/commits/11dc5a9fe83218df832396f364040bfd680dc0fa)

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: I'm on the fence between ignoring and wrapping a call in a try-catch on the caller side. The best practice is to have a `require` and if something goes south then a try-catch will inform about the error. I think this approach sort of "decouples" this contract and doesn't care who's going to call it. On the other hand, try-catch will add extra costs and at least for now, we know what contracts will call the `ReimbursmentPool` contract. Anyway, I removed the result of the `.call`, but also open for alternatives. [063786d](https://github.com/keep-network/keep-core/pull/2826/commits/063786d85371ffa16d1c170106990fd94acf1061)

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: Didn't work, solhint is still complaining.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: Simplified a bit [59b67ee](https://github.com/keep-network/keep-core/pull/2826/commits/59b67ee770f0031653bdc57cda01165da6c9e48c)

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: [3b8beed](https://github.com/keep-network/keep-core/pull/2826/commits/3b8beedcb07572f2c104749bb30ad0b6b16c11d1)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts:93]: part of [3169069](https://github.com/keep-network/keep-core/pull/2826/commits/3169069e0d8d41d3f953811ae1a6f6bc99e68e08)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts:171]: part of [3169069](https://github.com/keep-network/keep-core/pull/2826/commits/3169069e0d8d41d3f953811ae1a6f6bc99e68e08)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts:250]: part of [3169069](https://github.com/keep-network/keep-core/pull/2826/commits/3169069e0d8d41d3f953811ae1a6f6bc99e68e08)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: part of [3169069](https://github.com/keep-network/keep-core/pull/2826/commits/3169069e0d8d41d3f953811ae1a6f6bc99e68e08)

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:88]: part of [3169069](https://github.com/keep-network/keep-core/pull/2826/commits/3169069e0d8d41d3f953811ae1a6f6bc99e68e08)

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:132]: I'm not sure if there's much value to it because:
- we have `requires` that will tell if the transaction fails, especially this one: `require(sent, "Failed to send Ether")`
- it will be executed by the gov which can see right away if the funds were received at the provided address
- it adds extra cost

Unless you have a specific case in mind when it's needed?

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (nkuba) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts:250]: One more test case please: `refund` when the contract's balance is not high enough

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (nkuba) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:133]: It can be useful for maintainers/operators to know that some movement happened in the contract and there is a risk that transactions may not be reimbursable. 
`withdraw` is not a transaction that is expected to be executed often, so I don't think we should be worried about adding an event here.

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (nkuba) [inline solidity/random-beacon/test/ReimbursmentPool.test.ts:18]:
```suggestion
  let authorizedContract: SignerWithAddress
```

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (nkuba) [inline solidity/random-beacon/test/ReimbursmentPool.test.ts:287]: Let's extract this check to a separate `it` test case.

Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: [b2781e1](https://github.com/keep-network/keep-core/pull/2826/commits/b2781e16230fd4935ec19337a6b5cd5792a60526)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: Changed to a more general `thirdPartyContract`, cause `authorizedContract` also won't fit in the `authorize` tests, ie. it's not yet authorized. [b2781e1](https://github.com/keep-network/keep-core/pull/2826/commits/b2781e16230fd4935ec19337a6b5cd5792a60526)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: [b2781e1](https://github.com/keep-network/keep-core/pull/2826/commits/b2781e16230fd4935ec19337a6b5cd5792a60526)

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:135]: [b2781e1](https://github.com/keep-network/keep-core/pull/2826/commits/b2781e16230fd4935ec19337a6b5cd5792a60526)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (nkuba) [inline solidity/random-beacon/test/ReimbursmentPool.test.ts:422]: We should have a before block where we first authorize the address.

**Review by nkuba — CHANGES_REQUESTED** [review]

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: [c03c038](https://github.com/keep-network/keep-core/pull/2826/commits/c03c038461a772771d7ae100b372f9c81d46c230)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts:422]: good point [c03c038](https://github.com/keep-network/keep-core/pull/2826/commits/c03c038461a772771d7ae100b372f9c81d46c230)

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:42]: Why not `AuthorizeContractForRefund` or `DeauthorizeContractForRefund`? Alternatively just `AuthorizeContract` or `DeauthorizeContract`?

Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: Uh, that's clearly a global scope find-replace error, thx for catching it. They were named previously  `AuthorizeContract` and `UnauthorizeContract`. Changed in [28daf23](https://github.com/keep-network/keep-core/pull/2826/commits/28daf23db05be69b14c19e60676812fe7a831cf5)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts:8]: I did a `yarn format:fix` and haven't noticed any warnings, so not sure what are you referring to?

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: Removed

File: solidity/random-beacon/test/fixtures/index.ts
Comment (dimpar) [reply solidity/random-beacon/test/fixtures/index.ts]: Static gas cost includes the execution of code in the `refund()` + transaction cost. This number was captured at the moment of implementing the tests a while ago. After addressing some of the comments the avg. gas bumped to ~41k for the `refund()`. The static gas will be passed in the constructor during the deployment and should be probably revisited then.

```
gas for refund (min, max, avg) 40433 | 41893 | 41075  
```

I don't think we can precisely tell what is the cost of the `refund()`. We have min, max and avg gas usage depending on the situation. If we assume that the avg. is 41k then at most a caller will be net-negative around 800-900gas. Assuming the gas is 100gwei that is around 0,00008 ETH. Or can be a bit net-positive if a user hits a min gas path.

More integration tests between ECDSA Wallet registry / random-beacon and Reimbursement pool will be added in the next PRs where the refunds are checked.

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: changed

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: copy-pasta error, changed.

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: removed `beforeEach`

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: I've added more descriptions. The idea of the lower range at the moment of writing tests was to check if the gas is greater than 73000 assuming the gas price was 1gwei. It was just to show it's not 0, but I see that it might be a bit confusing, so I changed it just to `>0`. The higher range is calculated based on: `gas*tx.gasprice` which is a bit less than 145k. The precise number in the other test is based on the `maxGasPrice` which is set to 1gwei for testing purposes.

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (pdyraga) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts:8]: image

Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts:8]: Strange, can't recreate it on my local. I removed `node_modules, typechain, build, cache` folders and do the `yarn lint`. No warnings. I'm on node 14.9.0. @nkuba do you see the same warnings on your box?

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (nkuba) [inline solidity/random-beacon/test/ReimbursmentPool.test.ts:416]:
```suggestion
  describe("unauthorize", () => {
```

**Review by nkuba — CHANGES_REQUESTED** [review]

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:64]: What is the gas cost difference with/without non-reentrant? All approved contracts are supposed to be trusted.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/ReimbursmentPool.sol:42]: All other events are named using the past tense. Can we do the same for these two?

File: solidity/random-beacon/test/fixtures/index.ts
Comment (pdyraga) [reply solidity/random-beacon/test/fixtures/index.ts]: It would be good to establish this number before merging this PR. `ReimbursmentPool` code should not change in the future so there is no better moment than now, especially that as you say, we will have more tests added in the following PRs and we should be able to validate our choice this way.

I think the strategy should be: if we can't use a precise number, we establish one that is returning a tiny bit more than actually spent. If we see a frontrunning in the network, we will lower it and everyone will receive a tiny bit less than what they have actually spent.

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol:64]: With `nonReentrant`: 40433  ·      41893  ·      41075 
Without:  38083  ·      39543  ·      38725

If per prev. comments we take `max`, then the diff is `3168`. 

Yes, there are supposed to be trusted, but what if this trusted contract has a bug that can enable reentrancy here? The diff is not that small, but also not too big compared to the overall cost and having in mind the risk of draining the pool maybe we should leave the `nonReentrant` modifier in place?

File: solidity/random-beacon/contracts/ReimbursmentPool.sol
Comment (dimpar) [reply solidity/random-beacon/contracts/ReimbursmentPool.sol]: [ffab18c](https://github.com/keep-network/keep-core/pull/2826/commits/ffab18c52e13342a15d6195aa3c8dfd848c05da5)

File: solidity/random-beacon/test/ReimbursmentPool.test.ts
Comment (dimpar) [reply solidity/random-beacon/test/ReimbursmentPool.test.ts]: [cbc5623](https://github.com/keep-network/keep-core/pull/2826/commits/cbc5623a574ef8563b12801b3ead13da9b05b9e8)

**Review by pdyraga — APPROVED** [review]

**Review by nkuba — APPROVED** [review]

**Pull request merged**
